scipy/stats/tests/test_hypotests.py

```diff
@@ -164,6 +166,412 @@ def test_callable_cdf(self):
+
+    @pytest.mark.parametrize(("kwds", "result"), cases_scalar)
+    def test_scalar_data(self, kwds, result):
+        # just making sure scalars work
+        assert_allclose(mannwhitneyu(1, 2, **kwds), result)
+
+    def test_equal_scalar_data(self):
+        # when two scalars are equal, there is an -0.5/0 in the asymptotic
+        # approximation. R gives pvalue=1.0 for alternatives 'less' and
+        # 'greater' but NA for 'two-sided'. I don't see why, so I don't
+        # see a need for a special case to match that behavior.
+        assert_equal(mannwhitneyu(1, 1, method="exact"), (0.5, 1))
+        assert_equal(mannwhitneyu(1, 1, method="asymptotic"), (0.5, 1))
+
+        # without continuity correction, this becomes 0/0, which really
+        # is undefined
+        assert_equal(mannwhitneyu(1, 1, method="asymptotic",
+                                  use_continuity=False), (0.5, np.nan))
+
+    # --- Test Enhancements / Bug Reports ---
+
+    @pytest.mark.parametrize("method", ["asymptotic", "exact"])
+    def test_gh_12837_11113(self, method):
+        # Test that behavior for broadcastable nd arrays is appropriate:
+        # output shape is correct and all values are equal to when the test
+        # is performed on one pair of samples at a time.
+        # Tests that gh-12837 and gh-11113 (requests for n-d input)
+        # are resolved
+        np.random.seed(0)
+
+        # arrays are broadcastable except for axis = -3
+        axis = -3
+        m, n = 7, 10  # sample sizes
+        x = np.random.rand(m, 3, 8)
+        y = np.random.rand(6, n, 1, 8) + 0.1
+        res = mannwhitneyu(x, y, method=method, axis=axis)
+
+        shape = (6, 3, 8)  # appropriate shape of outputs, given inputs
+        assert(res.pvalue.shape == shape)
+        assert(res.statistic.shape == shape)
+
+        # move axis of test to end for simplicity
+        x, y = np.moveaxis(x, axis, -1), np.moveaxis(y, axis, -1)
+
+        x = x[None, ...]  # give x a zeroth dimension
+        assert(x.ndim == y.ndim)
+
+        x = np.broadcast_to(x, shape + (m,))
+        y = np.broadcast_to(y, shape + (n,))
+        assert(x.shape[:-1] == shape)
+        assert(y.shape[:-1] == shape)
+
+        # loop over pairs of samples
+        statistics = np.zeros(shape)
+        pvalues = np.zeros(shape)
+        for indices in product(*[range(i) for i in shape]):
+            xi = x[indices]
+            yi = y[indices]
+            temp = mannwhitneyu(xi, yi, method=method)
+            statistics[indices] = temp.statistic
+            pvalues[indices] = temp.pvalue
+
+        np.testing.assert_equal(res.pvalue, pvalues)
+        np.testing.assert_equal(res.statistic, statistics)
+
+    def test_gh_11355(self):
+        # Test for correct behavior with NaN/Inf in input
+        x = [1, 2, 3, 4]
+        y = [3, 6, 7, 8, 9, 3, 2, 1, 4, 4, 5]
+        res1 = mannwhitneyu(x, y)
+
+        # Inf is not a problem. This is a rank test, and it's the largest value
```

Comment:
Sure. Infs are ties, but it's not a problem in calculating the statistic. Added additional test against wilcox.R though for good measure.

Reply:
great, I saw it. if we need another commit, let's add that the test cases were generated with R's wilcox.test

Reply:
btw: I cannot reproduce the results in R, maybe I am not setting the parameters correctly? For example, for the last test case, I get a statistic of 4.5 instead of 24.5:

```
x <- c(1, Inf, Inf, 4)
y <- c(3, 6, 7, 8, Inf, Inf, 2, 1, 4, 4, 5)
wilcox.test(x, y) # W = 4.5, p-value = 0.3402
```
For the first test, I can reproduce the statistic in R but not the p-value no matter if I set correction=TRUE or FALSE.

Reply:
I run `wilcox.test` here: https://rdrr.io/r/stats/wilcox.test.html

```R
options(digits = 16)

x = c(1, 2, 3, 4)
y = c(3, 6, 7, 8, Inf, 3, 2, 1, 4, 4, 5)
wilcox.test(x, y, exact=FALSE)

x = c(1, 2, 3, 4)
y = c(3, 6, 7, 8, Inf, Inf, 2, 1, 4, 4, 5)
wilcox.test(x, y, exact=FALSE)

x = c(1, 2, Inf, 4)
y = c(3, 6, 7, 8, Inf, 3, 2, 1, 4, 4, 5)
wilcox.test(x, y, exact=FALSE)

x = c(1, 2, Inf, 4)
y = c(3, 6, 7, 8, Inf, Inf, 2, 1, 4, 4, 5)
wilcox.test(x, y, exact=FALSE)

x = c(1, Inf, Inf, 4)
y = c(3, 6, 7, 8, Inf, Inf, 2, 1, 4, 4, 5)
wilcox.test(x, y, exact=FALSE, alternative='t')

x <- c(1, Inf, Inf, 4)
y <- c(3, 6, 7, 8, Inf, Inf, 2, 1, 4, 4, 5)
wilcox.test(x, y) # cannot compute exact p-value with ties 
```

Reply:
OK, I tried with R version 4.0.5 (2021-03-31) and I can reproduce the values.